kroxylicious/src/main/java/io/kroxylicious/proxy/filter/multitenant/MultiTenantTransformationFilter.java

```diff
@@ -0,0 +1,233 @@
+        response.topics().forEach(topic -> removeTenantPrefix(context, topic::name, topic::setName, false));
+        response.groups().forEach(responseGroup -> responseGroup.topics().forEach(topic -> removeTenantPrefix(context, topic::name, topic::setName, false)));
+        context.forwardResponse(response);
+    }
+
+    @Override
+    public void onOffsetForLeaderEpochRequest(OffsetForLeaderEpochRequestData request, KrpcFilterContext context) {
+        request.topics().forEach(topic -> applyTenantPrefix(context, topic::topic, topic::setTopic, false));
+        context.forwardRequest(request);
+    }
+
+    @Override
+    public void onOffsetForLeaderEpochResponse(OffsetForLeaderEpochResponseData response, KrpcFilterContext context) {
+        response.topics().forEach(topic -> removeTenantPrefix(context, topic::topic, topic::setTopic, false));
+        context.forwardResponse(response);
+    }
+
+    @Override
+    public void onOffsetCommitRequest(OffsetCommitRequestData request, KrpcFilterContext context) {
+        request.topics().forEach(topic -> applyTenantPrefix(context, topic::name, topic::setName, false));
+        context.forwardRequest(request);
+    }
+
+    @Override
+    public void onOffsetCommitResponse(OffsetCommitResponseData response, KrpcFilterContext context) {
+        response.topics().forEach(topic -> removeTenantPrefix(context, topic::name, topic::setName, false));
+        context.forwardResponse(response);
+    }
+
+    @Override
+    public void onFetchRequest(FetchRequestData request, KrpcFilterContext context) {
+        request.topics().forEach(topic -> applyTenantPrefix(context, topic::topic, topic::setTopic, topic.topicId() != null));
+        context.forwardRequest(request);
+    }
+
+    @Override
+    public void onFetchResponse(FetchResponseData response, KrpcFilterContext context) {
+        response.responses().forEach(topic -> removeTenantPrefix(context, topic::topic, topic::setTopic, topic.topicId() != null));
+        context.forwardResponse(response);
+
+    }
+
+    private void applyTenantPrefix(KrpcFilterContext context, Supplier<String> getter, Consumer<String> setter, boolean ignoreEmpty) {
+        var tenantPrefix = getTenantPrefix(context);
+        String clientSideName = getter.get();
+        if (ignoreEmpty && (clientSideName == null || clientSideName.isEmpty())) {
+            return;
+        }
+        setter.accept(tenantPrefix + "-" + clientSideName);
+    }
+
+    private void removeTenantPrefix(KrpcFilterContext context, Supplier<String> getter, Consumer<String> setter, boolean ignoreEmpty) {
+        var tenantPrefix = getTenantPrefix(context);
+        var brokerSideName = getter.get();
+
+        if (ignoreEmpty && (brokerSideName == null || brokerSideName.isEmpty())) {
+            return;
+        }
+
+        setter.accept(brokerSideName.substring(tenantPrefix.length() + 1));
+    }
+
+    private static String getTenantPrefix(KrpcFilterContext context) {
```

Comment:
Why would another filter need to care about tenancy in general, or the concept of a tenant prefix in particular?

Here's an example: A transparent multitenancy filter included in the same chain as a topic encryption filter. Tenants need to have different encryption keys, so tenant1's `foo` (i.e. `tenant1.foo`) has `key1` and tenant2's `foo` (i.e. `tenant2.foo`) has `key2`. There _is_ some small convenience value in being able to ask the context for the tenant id. But I would argue that a topic encryption filter is _likely_ to want to support a prefix-based key lookup policy because Kafka's own (nontransparent) multitenancy support is based on using prefixes. So to be useful for a topic encryption filter to support that, because that way it works exactly the same when used in a chain with, or without, the MT filter. So it's really not clear to me that the convenience value is worth the additional complexity of forcing all filters to understand multitenancy (i.e. decide whether they need to use the tenant id from the context, document whether they use it, and how it changes the semantics of the filter). 

For me conceptual simplicity, composability and uniformity of behaviour are really important because they make it simpler for people writing filters (just focus on what your filter needs to do in terms of the Kafka protocol, you don't have to worry about concerns being forced on to you from Kroxy itself), and for people using filters (just combine the filters you want to use), including 3rd party filters (i.e. this these principles could help support an ecosystem).  Yes, that can sometimes make it slightly awkward if the operator needs to extract the tenant id, but I think conventions about filters supporting prefix-based policies/configurations might be sufficient.